src/game/ingame_menu.c

```diff
@@ -224,15 +208,160 @@ void create_dl_ortho_matrix(void) {
 }
 
-// Unused
-#if FALSE
+/**
+ * Determine which UTF-8 character to render, given a string and the current position in the string.
+ * Returns the table entry of the relevant character.
+ * Also increments the string position by the correct amount to reach the next character.
+ */
+struct Utf8CharLUTEntry *utf8_lookup(struct Utf8LUT *lut, char *str, s32 *strPos) {
+    u32 codepoint;
+    struct Utf8CharLUTEntry *usedLUT;
+    u32 length;
+
+    lut = segmented_to_virtual(lut);
+    if (!(str[*strPos] & 0x20)) {
+        codepoint = ((str[*strPos] & 0x1F) << 6) | (str[*strPos + 1] & 0x3F);
+        *strPos += 1;
+
+        usedLUT = segmented_to_virtual(lut->lut2Bytes);
+        length = lut->length2Bytes;
+    } else if (!(str[*strPos] & 0x10)) {
+        codepoint = ((str[*strPos] & 0xF) << 12) | ((str[*strPos + 1] & 0x3F) << 6) | (str[*strPos + 2] & 0x3F);
+        *strPos += 2;
+
+        usedLUT = segmented_to_virtual(lut->lut3Bytes);
+        length = lut->length3Bytes;
+    } else {
+        codepoint = ((str[*strPos] & 0x7) << 18) | ((str[*strPos + 1] & 0x3F) << 12) | ((str[*strPos + 2] & 0x3F) << 6) | (str[*strPos + 3] & 0x3F);
+        *strPos += 3;
+
+        usedLUT = segmented_to_virtual(lut->lut4Bytes);
+        length = lut->length4Bytes;
+    }
+
+    s32 start = 0;
+    s32 end = length - 1;
+    s32 mid = (start + end) / 2;
+
+    while (start <= end) {
+        if (usedLUT[mid].codepoint == codepoint) {
+            return &usedLUT[mid];
+        }
+
+        if (usedLUT[mid].codepoint > codepoint) {
+            end = mid - 1;
+        } else {
+            start = mid + 1;
+        }
+
+        mid = (start + end) / 2;
+    }
+    return segmented_to_virtual(lut->missingChar);
+}
+
+/**
+ * Convert a character in the range 0-9 or A-F to an integer value. Returns -1 if the character is
+ * not a valid hex digit.
+ */
+static s32 hex_char_to_value(char c) {
+    if (c >= '0' && c <= '9') {
+        return c - '0';
+    }
+    if (c >= 'A' && c <= 'F') {
+        return c - 'A' + 10;
+    }
+    if (c >= 'a' && c <= 'f') {
+        return c - 'a' + 10;
+    }
+    return -1;
+}
+
+/**
+ * Determine if the characters following an @ sign are a valid hex color code.
+ */
+static s32 is_color_code_valid(char *str, s32 strPos) {
+    for (s32 i = 0; i < 6; i++) {
+        if (hex_char_to_value(str[strPos + i]) == -1) {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+/**
+ * Get the exact length of a string of any font in pixels, using the given ASCII and UTF-8 tables.
+ */
+s32 get_string_length(char *str, struct AsciiCharLUTEntry *asciiLut, struct Utf8LUT *utf8LUT) {
+    s32 length = 0;
+    s32 strPos = 0;
+
+    s32 checkColorCodes = (asciiLut == main_font_lut); // Only check for @ if the font is the main generic font.
+
+    asciiLut = segmented_to_virtual(asciiLut);
+
+    while (str[strPos] != '\0') {
+        if (checkColorCodes && str[strPos] == '@') {
```

Comment:
%s and %d are only really meant to be used by dialog and you would basically never use this function on a dialog string. my main concern is that it would be bad for performance to have extra checks

Reply:
coming back to this i'd like to mention that there is basically no reason to use %d and %s OUTSIDE of dialog since you can use sprintf directly. the only reason i have it like this for dialog is that you would need a HUGE buffer to do it with sprintf and you barely need it most of the time anyway. for normal strings a buffer and sprintf is way easier